crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+            boundaries: Vec::with_capacity(iter.len() + 1),
+        };
+        out.boundaries.push(0);
+        for (label, coeff) in iter {
+            out.add_dense_label(&label, coeff)?;
+        }
+        Ok(out)
+    }
+
+    /// Construct an observable from a list of labels, the qubits each item applies to, and the
+    /// coefficient of the whole term.
+    ///
+    /// This is analagous to :meth:`.SparsePauliOp.from_sparse_list`, except it uses
+    /// :ref:`the extended alphabet <sparse-observable-alphabet>` of :class:`.SparseObservable`.
+    ///
+    /// The "labels" and "indices" fields of the triples are associated by zipping them together.
+    /// For example, this means that a call to :meth:`from_list` can be converted to the form used
+    /// by this method by setting the "indices" field of each triple to ``(num_qubits-1, ..., 1,
+    /// 0)``.
+    ///
+    /// Args:
+    ///     iter (list[tuple[str, Sequence[int], complex]]): triples of labels, the qubits
+    ///         each single-qubit term applies to, and the coefficient of the entire term.
+    ///
+    ///     num_qubits (int): the number of qubits in the operator.
+    ///
+    /// Examples:
+    ///
+    ///     Construct a simple operator::
+    ///
+    ///         >>> SparseObservable.from_sparse_list(
+    ///         ...     [("ZX", (1, 4), 1.0), ("YY", (0, 3), 2j)],
+    ///         ...     num_qubits=5,
+    ///         ... )
+    ///         <SparseObservable with 2 terms on 5 qubits: (1+0j)(X_4 Z_1) + (0+2j)(Y_3 Y_0)>
+    ///
+    ///     Construct the identity observable (though really, just use :meth:`identity`)::
+    ///
+    ///         >>> SparseObservable.from_sparse_list([("", (), 1.0)], num_qubits=100)
+    ///         <SparseObservable with 1 term on 100 qubits: (1+0j)()>
+    ///
+    ///     This method can replicate the behavior of :meth:`from_list`, if the qubit-arguments
+    ///     field of the triple is set to decreasing integers::
+    ///
+    ///         >>> labels = ["XY+Z", "rl01", "-lXZ"]
+    ///         >>> coeffs = [1.5j, 2.0, -0.5]
+    ///         >>> from_list = SparseObservable.from_list(list(zip(labels, coeffs)))
+    ///         >>> from_sparse_list = SparseObservable.from_sparse_list([
+    ///         ...     (label, (3, 2, 1, 0), coeff)
+    ///         ...     for label, coeff in zip(labels, coeffs)
+    ///         ... ])
+    ///         >>> assert from_list == from_sparse_list
+    #[staticmethod]
+    #[pyo3(name = "from_sparse_list", signature = (iter, /, num_qubits))]
+    fn py_from_sparse_list(
+        iter: Vec<(String, Vec<u32>, Complex64)>,
+        num_qubits: u32,
+    ) -> Result<Self, LabelError> {
+        let coeffs = iter.iter().map(|(_, _, coeff)| *coeff).collect();
+        let mut boundaries = Vec::with_capacity(iter.len() + 1);
+        boundaries.push(0);
+        let mut indices = Vec::new();
+        let mut bit_terms = Vec::new();
+        let mut sorted = btree_map::BTreeMap::new();
```

Comment:
This is to ensure the indices are sorted in ascending order, right? It would be nice to add a comment for our future selves 🙂

Reply:
Yeah, a core purpose of a `BTreeMap` is that its iteration methods all return their results in sorted order.  It modifies the asymptotic complexities of a few actions away from a "pure" hashmap - it's kind of a pragmatic hybrid of a binary tree, a hashmap and a list.

Reply:
More comment added in 9dd0187213.